tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc

```diff
@@ -312,17 +312,70 @@ llvm::Value* EmitPrintf(absl::string_view fmt,
        arguments_ptr});
 }
 
+// Helper function to emit call to AMDGPU shfl_down function.
+llvm::Value* EmitShflDownDeviceFunctionForAMDGPU(
+    absl::Span<llvm::Value* const> operands, llvm::IRBuilder<>* b) {
+  llvm::Module* module = b->GetInsertBlock()->getModule();
+  // AMDGPU device function requires first argument - value - as i32.
+  llvm::Value* value = b->CreateBitCast(operands[0], b->getInt32Ty());
+  llvm::Value* offset = operands[1];
+  // The input type is {i32, i32}.
+  std::vector<llvm::Type*> ir_input_types(2, b->getInt32Ty());
+  // The output type is i32.
+  llvm::Type* ir_output_type = b->getInt32Ty();
+  llvm::FunctionType* callee_type =
+      llvm::FunctionType::get(/*Result=*/ir_output_type, ir_input_types,
+                              /*isVarArg=*/false);
+  string callee_name = "__ockl_readuplane_i32";
+  llvm::FunctionCallee shfl_call = module->getOrInsertFunction(
+      llvm_ir::AsStringRef(callee_name), callee_type);
+  llvm::Value* result = b->CreateCall(shfl_call, {value, offset});
+  int bit_width = value->getType()->getPrimitiveSizeInBits();
+  if (operands[0]->getType()->isFloatTy() && bit_width == 32) {
+    return b->CreateBitCast(result,
```

Comment:
Just like above we do the bitcast unconditionally, can we do it unconditionally here?

Reply:
For the case when value is not f32 the  call site creates a  i32 vector and does an insertelement of the returned value.  So it expects an i32. So thats why I kept the conditional cast to float.

Reply:
Sorry, I do not understand.

I was suggesting that we change this if/else statement to simply:

```
return b->CreateBitCast(result, operands[0]->getType());
```

I thought this would not change the behavior of this function.

I see that the caller does an insertelement, but I do not understand how that affects anything.

Reply:
Sorry..I misunderstood..have made it an unconditional bitcast